src/core/pointcloud/qgspointcloudlayerexporter.cpp

```diff
@@ -0,0 +1,585 @@
+    saveOptions.actionOnExistingFile = mActionOnExistingFile;
+    saveOptions.feedback = mFeedback;
+    mVectorSink = QgsVectorFileWriter::create( mFilename, outputFields(), QgsWkbTypes::PointZ, mTargetCrs, QgsCoordinateTransformContext(), saveOptions );
+    ExporterVector exp = ExporterVector( this );
+    exp.run();
+  }
+}
+
+QgsMapLayer *QgsPointCloudLayerExporter::getExportedLayer()
+{
+  if ( mFormat == QLatin1String( "memory" ) && mMemoryLayer )
+  {
+    QgsMapLayer *retVal = mMemoryLayer;
+    mMemoryLayer = nullptr;
+    return retVal;
+  }
+
+  const QFileInfo fileInfo( mFilename );
+
+  if ( mFormat == QLatin1String( "LAZ" ) )
+  {
+    return new QgsPointCloudLayer( mFilename, fileInfo.completeBaseName(), QStringLiteral( "pdal" ) );
+  }
+
+  if ( mFormat == QLatin1String( "GPKG" ) )
+  {
+    QString uri( mFilename );
+    uri += "|layername=" + mName;
+    return new QgsVectorLayer( uri, mName, QStringLiteral( "ogr" ) );
+  }
+
+  return new QgsVectorLayer( mFilename, fileInfo.completeBaseName(), QStringLiteral( "ogr" ) );
+}
+
+//
+// ExporterBase
+//
+
+void QgsPointCloudLayerExporter::ExporterBase::run()
+{
+  QVector<IndexedPointCloudNode> nodes;
+  qint64 pointCount = 0;
+  QQueue<IndexedPointCloudNode> queue;
+  queue.push_back( mParent->mIndex->root() );
+  while ( !queue.empty() )
+  {
+    IndexedPointCloudNode node = queue.front();
+    queue.pop_front();
+    if ( !nodes.contains( node ) &&
+         mParent->mExtent.intersects( mParent->mIndex->nodeMapExtent( node ) ) &&
+         mParent->mZRange.overlaps( mParent->mIndex->nodeZRange( node ) ) )
+    {
+      pointCount += mParent->mIndex->nodePointCount( node );
+      nodes.push_back( node );
+    }
+    for ( const IndexedPointCloudNode &child : mParent->mIndex->nodeChildren( node ) )
+    {
+      queue.push_back( child );
+    }
+  }
+
+
+
+  int pointsSkipped = 0;
+  const qint64 pointsToExport = std::max< qint64 >( std::min( mParent->mPointsLimit, pointCount ), 1 );
+  QgsPointCloudRequest request;
+  request.setAttributes( mParent->requestedAttributeCollection() );
+  std::unique_ptr<QgsPointCloudBlock> block = nullptr;
+  qint64 pointsExported = 0;
+  for ( const IndexedPointCloudNode &node : nodes )
+  {
+    block.reset( mParent->mIndex->nodeData( node, request ) );
+    const QgsPointCloudAttributeCollection attributesCollection = block->attributes();
+    const char *ptr = block->data();
+    int count = block->pointCount();
+    int recordSize = attributesCollection.pointRecordSize();
+    const QgsVector3D scale = block->scale();
+    const QgsVector3D offset = block->offset();
+    int xOffset;
+    int yOffset;
+    int zOffset;
+    attributesCollection.find( QStringLiteral( "X" ), xOffset );
+    attributesCollection.find( QStringLiteral( "Y" ), yOffset );
+    attributesCollection.find( QStringLiteral( "Z" ), zOffset );
+    for ( int i = 0; i < count; ++i )
+    {
+
+      if ( mParent->mFeedback )
```

Comment:
I would suggest to update feedback only once upon a time, not with every point... e.g. `if ( i % 1000 == 0 )`

Reply:
Whoops!  I overlooked this! Cancelling will still be snappy if checked every 1000 points. Done